x-pack/plugins/observability_solution/synthetics/server/routes/overview_status/overview_status_service.ts

```diff
@@ -0,0 +1,331 @@
+      } while (hasMoreData && afterKey);
+      return monitorByIds;
+    });
+  }
+
+  processOverviewStatus(
+    monitors: Array<SavedObjectsFindResult<EncryptedSyntheticsMonitorAttributes>>,
+    statusData: Map<string, LocationStatus>
+  ) {
+    let up = 0;
+    let down = 0;
+    const upConfigs: Record<string, OverviewStatusMetaData> = {};
+    const downConfigs: Record<string, OverviewStatusMetaData> = {};
+    const pendingConfigs: Record<string, OverviewStatusMetaData> = {};
+    const disabledConfigs: Record<string, OverviewStatusMetaData> = {};
+
+    const enabledMonitors = monitors.filter((monitor) => monitor.attributes[ConfigKey.ENABLED]);
+    const disabledMonitors = monitors.filter((monitor) => !monitor.attributes[ConfigKey.ENABLED]);
+
+    const queryLocIds = this.filterData?.locationFilter;
+
+    disabledMonitors.forEach((monitor) => {
+      const monitorQueryId = monitor.attributes[ConfigKey.MONITOR_QUERY_ID];
+      const meta = this.getMonitorMeta(monitor);
+      monitor.attributes[ConfigKey.LOCATIONS]?.forEach((location) => {
+        disabledConfigs[`${meta.configId}-${location.id}`] = {
+          monitorQueryId,
+          status: 'disabled',
+          locationId: location.id,
+          locationLabel: location.label,
+          ...meta,
+        };
+      });
+    });
+
+    enabledMonitors.forEach((monitor) => {
+      const monitorId = monitor.attributes[ConfigKey.MONITOR_QUERY_ID];
+      const monitorStatus = statusData.get(monitorId);
+
+      // discard any locations that are not in the monitorLocationsMap for the given monitor as well as those which are
+      // in monitorLocationsMap but not in listOfLocations
+      const monLocations = monitor.attributes[ConfigKey.LOCATIONS];
+      // const monQueriedLocations = intersection(monLocations, monitorLocationIds);
+      monLocations?.forEach((monLocation) => {
+        if (!isEmpty(queryLocIds) && !queryLocIds?.includes(monLocation.id)) {
+          // filter out location provided via query
+          return;
+        }
+        const locData = monitorStatus?.find((loc) => loc.locationId === monLocation.id);
+        const meta = {
+          monitorQueryId: monitorId,
+          locationId: monLocation.id,
+          timestamp: locData?.timestamp,
+          locationLabel: monLocation.label,
+          ...this.getMonitorMeta(monitor),
+        };
+        const monLocId = `${meta.configId}-${monLocation.id}`;
+        if (locData) {
+          if (locData.status === 'down') {
+            down += 1;
+            downConfigs[monLocId] = {
+              ...meta,
+              status: 'down',
+            };
+          } else if (locData.status === 'up') {
+            up += 1;
+            upConfigs[monLocId] = {
+              ...meta,
+              status: 'up',
+            };
+          }
+        } else {
+          pendingConfigs[monLocId] = {
+            status: 'unknown',
+            ...meta,
+          };
+        }
+      });
+    });
```

Comment:
```suggestion
    const queryLocIds = this.filterData?.locationFilter;
    for (const monitor of monitors) {
      if (!monitor.attributes[ConfigKey.ENABLED]) {
        const monitorQueryId = monitor.attributes[ConfigKey.MONITOR_QUERY_ID];
        const meta = this.getMonitorMeta(monitor);
        monitor.attributes[ConfigKey.LOCATIONS]?.forEach((location) => {
          disabledConfigs[`${meta.configId}-${location.id}`] = {
            monitorQueryId,
            status: 'disabled',
            locationId: location.id,
            locationLabel: location.label,
            ...meta,
          };
        });
      } else {
        const monitorId = monitor.attributes[ConfigKey.MONITOR_QUERY_ID];
        const monitorStatus = statusData.get(monitorId);

        // discard any locations that are not in the monitorLocationsMap for the given monitor as well as those which are
        // in monitorLocationsMap but not in listOfLocations
        const monLocations = monitor.attributes[ConfigKey.LOCATIONS];
        monLocations?.forEach((monLocation) => {
          if (!isEmpty(queryLocIds) && !queryLocIds?.includes(monLocation.id)) {
            // filter out location provided via query
            return;
          }
          const locData = monitorStatus?.find((loc) => loc.locationId === monLocation.id);
          const meta = {
            monitorQueryId: monitorId,
            locationId: monLocation.id,
            timestamp: locData?.timestamp,
            locationLabel: monLocation.label,
            ...this.getMonitorMeta(monitor),
          };
          const monLocId = `${meta.configId}-${monLocation.id}`;
          if (locData) {
            if (locData.status === 'down') {
              down += 1;
              downConfigs[monLocId] = {
                ...meta,
                status: 'down',
              };
            } else if (locData.status === 'up') {
              up += 1;
              upConfigs[monLocId] = {
                ...meta,
                status: 'up',
              };
            }
          } else {
            pendingConfigs[monLocId] = {
              status: 'unknown',
              ...meta,
            };
          }
        });
      }
    }
```

We can reduce the number of loops here and it's a bit simpler. This changes it from 2 loops over the whole set and 2 loops over the subsets we generate, to a single loop over the whole set.

I ran this code against your unit tests and everything came up green.